### ISSUE TITLE
Save inconsistent knowledge

### ISSUE THREAD

**Issue by xtuchyna** [thoth-station/mi#371]
**Is your feature request related to a problem? Please describe.**
Currently, whenever the `mi` checks the data consistency of repository after extraction, if it finds the inconsistent data according to entity schema (let's say e.g. PR body is `None`, but in the scheme it was specify to only `str` and not `Any(str, None)` ) it cancels the data aggregation and everything stored is thrown away.

**Describe the solution you'd like**
Store data with `_inconsistent` suffix rather then not saving it at all. This can happen during analysing large repositories, where it is common that various entities can occur in edge cases (with e.g. `None` as a PR body).

**Comment by xtuchyna** [comment]
WDYT @pacospace , I think this is a bit more practical. Additionally, we could then write tests in `hypothesis` (#326 ) to ensure data consistency.

**Comment by pacospace** [comment]
> **Is your feature request related to a problem? Please describe.**
> Currently, whenever the `mi` checks the data consistency of repository after extraction, if it finds the inconsistent data according to entity schema (let's say e.g. PR body is `None`, but in the scheme it was specify to only `str` and not `Any(str, None)` ) it cancels the data aggregation and everything stored is thrown away.
> 
> **Describe the solution you'd like**
> Store data with `_inconsistent` suffix rather then not saving it at all. This can happen during analysing large repositories, where it is common that various entities can occur in edge cases (with e.g. `None` as a PR body).

- what do you mean all thrown away? I expect only that specific PR is thrown away not everything right?

- what do you do with those data? improve data aggregation of edge cases?

**Comment by xtuchyna** [comment]
> what do you mean all thrown away? I expect only that specific PR is thrown away not everything right?

That is reasonable, I can make a Feature request out of this then.

> what do you do with those data? improve data aggregation of edge cases?

I am not sure if I understand this

**Comment by pacospace** [comment]
> > what do you mean all thrown away? I expect only that specific PR is thrown away not everything right?
> 
> That is reasonable, I can make a Feature request out of this then.
> 
> > what do you do with those data? improve data aggregation of edge cases?
> 
> I am not sure if I understand this

my question is why `Store data with _inconsistent suffix rather then not saving it at all`, what do you with stored data?

**Comment by xtuchyna** [comment]
Aah, it can be useful for a data scientist to further work with data - do an exploratory analysis, look and see what's wrong, clean it and make it consistent.

**Comment by pacospace** [comment]
> Aah, it can be useful for a data scientist to further work with data - do an exploratory analysis, look and see what's wrong, clean it and make it consistent.

Do you have examples of these edge cases?

**Comment by xtuchyna** [comment]
Yes, e.g. https://github.com/thoth-station/mi/pull/351
tl;dr - the issue body could be null and that was (from my point of view) an edge case.

**Comment by pacospace** [comment]
> Yes, e.g. #351
> tl;dr - the issue body could be null and that was (from my point of view) an edge case.

Well you should look at the GitHub API for the version of pygithub you are considering and that API with the schema checks you have should match.

I don't consider these edge cases, rather missing consistency checks respect to the GitHub API. 

The PR or Issue structure from github API should match the version you are considering for the data, otherwise these "edge cases" can appear always. So the use of testing should be the solution here actually, rather then collecting data not required for the analysis.

**Comment by pacospace** [comment]
> Yes, e.g. #351
> tl;dr - the issue body could be null and that was (from my point of view) an edge case.

Why are you using `Any(None, str)` if Any considers any type. If you expect None or str only, you should use `Optional[str]`

**Comment by xtuchyna** [comment]
> > Yes, e.g. #351
> > tl;dr - the issue body could be null and that was (from my point of view) an edge case.
> 
> Why are you using `Any(None, str)` if Any considers any type. If you expect None or str only, you should use `Optional[str]` right?

the `voluptuous` module uses Any as an Optional (in contrast to `typing`)
